File: ESP32 (C++)/ESP_UHF/Session.cpp
#include "Session.h"

String epc;
std::tuple<int, int> values;
int last_seen;
int last_sent;

void Session::renew_token() {
  std::map<String, String> new_tokens;
  String access_tokens;

  std::tie(new_tokens, access_tokens) = get_access_token(tokens);

  if (new_tokens.empty()) {
    Serial.print("Fail to refresh tokens!");
    return;
  } else {
    tokens = new_tokens;
    access_token = access_token;
  }
}

void Session::expiredAndInfo() {
  int current_time = millis()/1000;
  std::list<String> expired_tags;

    for (const auto& entry : table) {
      const String& key = entry.first;
      const auto& tuple = entry.second;
      int last_seen = std::get<0>(tuple);
      int last_sent = std::get<0>(tuple);

      if (last_seen + clear_interval < current_time) {
          expired_tags.push_back(key);
          Serial.print("------- DELETED: ");
          Serial.println(key);
          continue;
      }

      Serial.print(key);
      Serial.print(" ");
      Serial.print(last_seen);
      Serial.print(" ");
      Serial.println(last_sent);
  }

  // Remove expired items from the table
  for (const auto &expired : expired_tags) {
      table.erase(expired);
  }
}

void Session::send_epcs() {
  for (String epc : incoming_tags) {
    int posRepsonse = postPosTransaction(access_token, epc);
  }

  //thread_info.onRun(expiredAndInfo);
  //thread_info.setInterval(0);
}

void Session::process_epcs() {
  int last_seen;
  int last_sent;

  if (!tokens.empty()) {
    if (!incoming_tags.empty()) {
      std::list<String> epcs = incoming_tags;
      std::map<String, std::tuple<int, int>> registry = table;
      int current_time = millis()/1000;

      for (String epc : epcs) {

        if (registry.find(epc) == registry.end()) {

          registry["epc"] = std::make_tuple(current_time, current_time);

        } else {

          std::tie(last_seen, last_sent) = registry["epc"];

          if (last_sent + debounce_interval < current_time) {
            last_sent = current_time;
            Serial.println("---------------SENT");
          }

          last_seen = current_time;
          registry["epc"] = std::make_tuple(last_seen, last_sent);
        }
      }

      Serial.println();Serial.println();

      incoming_tags = std::list<String>();
      table = registry;
    }
  }
}

File: LLRP (Python)/session.py
from api import get_tokens, get_access_token, postPosTransaction, handlePosApiResponse
from time import sleep, time
import threading

#-----------------------------------------------------------------------------------------

class Session():

    def __init__(self, reader, request_interval = 5, debounce_interval = 10, clear_interval = 20):  
        self.reader = reader
        self.request_interval = request_interval
        self.debounce_interval = debounce_interval
        self.clear_interval = clear_interval

        self.initiate()

    def initiate(self):        
        print('Login to API...')
        self.tokens = get_tokens()
        self.access_token = ''

        self.incoming_tags = []
        self.table = {}

        threading.Thread(target=self.receive_epcs, args=(), daemon=True).start()

    def renewToken(self):
        new_tokens, access_token = get_access_token(self.tokens)
        if not new_tokens:
            print('Fail to refresh tokens!')
            return
        else:
            self.tokens = new_tokens
            self.access_token = access_token

    def receive_epcs(self): #Gets read epcs
        start = time()
        while start + int(self.request_interval) > time():
            sleep(0.05)

        self.incoming_tags = self.reader.publishEvent()  

        threading.Thread(target=self.expiredAndInfo, args=(), daemon=True).start()
        threading.Thread(target=self.receive_epcs, args=(), daemon=True).start()

    def send(self, epc): #Send to API
        posResponse = postPosTransaction(self.access_token, epc)
        if posResponse:
            handlePosApiResponse(posResponse)
            posResponse.close()
            del posResponse

    def expiredAndInfo(self): #Prints info and manages expired 
    
        current_time = time()
        expired_tags = []
        print("{:<10} {:<10} {:<10}".format('EPC', '              LAST SEEN', '        LAST SENT'))

        for key, values in self.table.items():
            last_seen, last_sent = values

            if last_seen + self.clear_interval < current_time:
                expired_tags.append(key)
                print('------- DELETED: ', key)
                continue
                
            print("{:x<7s} {:x<7f} {:x<7f}".format(key, last_seen, last_sent))
            
        for expired in expired_tags: del self.table[expired]